std/socket.zig

```diff
@@ -0,0 +1,678 @@
+            SocketType.Raw => system.SOCK_RAW,
+            SocketType.SeqPacket => system.SOCK_SEQPACKET,
+        });
+    }
+};
+
+pub const Protocol = enum {
+    Unix,
+    TCP,
+    UDP,
+    IP,
+    IPV6,
+    RAW,
+    ICMP,
+
+    fn toInt(self: Protocol, comptime T: type) !T {
+        return @intCast(T, switch (self) {
+            Protocol.Unix => if (is_posix) 0 else return SocketError.ProtocolNotSupported,
+            Protocol.TCP => if (is_posix) system.PROTO_tcp else system.IPPROTO_TCP,
+            Protocol.UDP => if (is_posix) system.PROTO_udp else system.IPPROTO_UDP,
+            Protocol.IP => if (is_posix) system.PROTO_ip else system.IPPROTO_IP,
+            Protocol.IPV6 => if (is_posix) system.PROTO_ipv6 else system.IPPROTO_IPV6,
+            Protocol.RAW => if (is_posix) system.PROTO_raw else system.IPPROTO_RAW,
+            Protocol.ICMP => if (is_posix) system.PROTO_icmp else system.IPPROTO_ICMP,
+        });
+    }
+};
+
+pub const Shutdown = enum {
+    Read = if (is_posix) system.SHUT_RD else system.SD_RECEIVE,
+    Write = if (is_posix) system.SHUT_WR else system.SD_SEND,
+    Both = if (is_posix) system.SHUT_RDWR else system.SD_BOTH,
+
+    fn toInt(self: Shutdown, comptime T: type) T {
+        return @intCast(T, switch (self) {
+            Shutdown.Read => if (is_posix) system.SHUT_RD else system.SD_RECEIVE,
+            Shutdown.Write => if (is_posix) system.SHUT_WR else system.SD_SEND,
+            Shutdown.Both => if (is_posix) system.SHUT_RDWR else system.SD_BOTH,
+        });
+    }
+};
+
+// pub const Level = enum(u32) {
+//     IP = if (is_posix) system.SOL_IP ,
+//     IPV6 = if (is_posix) system.SOL_IPV6 ,
+//     RM = if (is_posix) system. ,
+//     TCP = if (is_posix) system. ,
+//     UDP = if (is_posix) system. ,
+//     IPX = if (is_posix) system. ,
+//     AppleTalk = if (is_posix) system. ,
+//     IRLMP = if (is_posix) system. ,
+//     Socket = if (is_posix) system. ,
+// };
+
+fn accept4Windows(fd: SocketFd, addr: ?*net.OsAddress, addrlen: ?*system.socklen_t, flags: u32) SocketFd {
+    const result = system.accept(fd, addr, @ptrCast(*c_int, addrlen));
+    const ioctl_mode = @boolToInt((@intCast(c_ulong, flags) & system.FIONBIO) == system.FIONBIO);
+    const no_inherit = (flags & system.WSA_FLAG_NO_HANDLE_INHERIT);
+
+    if (system.ioctlsocket(result, system.FIONBIO, &@intCast(c_ulong, ioctl_mode)) != 0) {
+        return InvalidSocketFd;
+    }
+
+    if (no_inherit == system.WSA_FLAG_NO_HANDLE_INHERIT) {
+        const handle = @intToPtr(system.HANDLE, result);
+        if (system.SetHandleInformation(handle, system.HANDLE_FLAG_INHERIT, 0) == system.FALSE) {
+            return InvalidSocketFd;
+        }
+    }
+
+    return result;
+}
+
+// add timeouts
+
+pub const Socket = struct {
+    fd: SocketFd,
+    domain: Domain,
+    socket_type: SocketType,
+    protocol: Protocol,
+
+    fn initSocket(fd: SocketFd, domain: Domain, socket_type: SocketType, protocol: Protocol) Socket {
```

Comment:
Shouldn't this be public so that a user can pass in an fd that's already open?

note that you can get the domain/socket_type/protocol via `getsockopt` `SO_DOMAIN`/`SO_TYPE`/`SO_PROTOCOL`